vue-components/src/components/Button.vue

```diff
@@ -19,19 +22,49 @@ import Vue from 'vue';
  * https://bugzilla.mozilla.org/show_bug.cgi?id=1581369#c5
  */
 /* eslint-enable no-trailing-spaces */
-export default Vue.extend( {
+export default defineComponent( {
 	name: 'Button',
 	props: {
 		/**
 		 * The type of the button
+		 *
+		 * Allowed values: `neutral`, `progressive`, `destructive`
 		 */
 		type: {
 			type: String,
 			validator( value: string ): boolean {
-				return [ 'neutral', 'primaryProgressive', 'primaryDestructive' ].includes( value );
+				return [ 'neutral', 'progressive', 'destructive' ].includes( value );
 			},
 			default: 'neutral',
 		},
+		/**
+		 * The variant of the button
+		 *
+		 * Allowed values: `normal`, `primary`
+		 */
+		variant: {
+			type: String,
+			validator( value: string ): boolean {
+				return [ 'normal', 'primary' ].includes( value );
+			},
+			default: 'normal',
+		},
+	},
+	setup( props: {
+		type: 'neutral'|'progressive'|'destructive';
+		variant: 'normal'|'primary';
+	} ) {
+		onMounted( () => {
+			const supportedCombinations = {
+				normal: [ 'neutral' ],
+				primary: [ 'progressive', 'destructive' ],
+			};
+			if ( !supportedCombinations[ props.variant ].includes( props.type ) ) {
+				throw new Error(
+					`The combination of variant "${props.variant}" and type "${props.type}" is not yet supported!`,
+				);
+			}
+		} );
 	},
```

Comment:
I was hoping this could move into the prop validation but apparently it's not possible to access a prop in another prop's validator. Is there a benefit to doing this in `setup`? Why not just put it in `mounted`?

```suggestion
	mounted() {
		const supportedCombinations = {
			normal: [ 'neutral' ],
			primary: [ 'progressive', 'destructive' ],
		};
		if ( !supportedCombinations[ this.variant ].includes( this.type ) ) {
			throw new Error(
				`The combination of variant "${this.variant}" and type "${this.type}" is not yet supported!`,
			);
		}
	},
```

Reply:
That was my first attempt as well^^

I tried to explain it in the commit message. For me TypeScript is complaining that the string in `this.variant` does not index the object `supportedCombinations` and therefore `supportedCombinations[ this.variant ]` has type any and that is disallowed.

And since one of the explicitly stated benefits/purposes of the setup function is better type support I decided to use that instead of trying to weirdly typecast things somehow or use `// ts-disable`.

But maybe I did something slightly wrong and your solution works with TypeScript? Sometimes it is strange what causes type errors.

Reply:
I thought it could be managed with a `PropType` on `variant` and `type` and a corresponding index type on `supportedCombinations`, but I also struggled with getting TS to properly recognize the `PropType` and gave up. Strange, I thought I've seen a similar case somewhere.